### PR TITLE
[net] apply clippy lints

### PR DESCRIPTION
**Pull request by wcampbell0x2a**
Applied helpful clippy lints to the network std library module.

### CODE REVIEW

File: library/std/src/net/ip.rs [hunk 1]
```diff
@@ -1262,10 +1259,10 @@ impl Ipv6Addr {
...
     #[rustc_const_unstable(feature = "const_ipv6", issue = "76205")]
     pub const fn is_unicast_link_local_strict(&self) -> bool {
-        (self.segments()[0] & 0xffff) == 0xfe80
-            && (self.segments()[1] & 0xffff) == 0
-            && (self.segments()[2] & 0xffff) == 0
-            && (self.segments()[3] & 0xffff) == 0
+        (self.segments()[0]) == 0xfe80
+            && self.segments()[1] == 0
+            && self.segments()[2] == 0
+            && self.segments()[3] == 0
```
Comment (LingMan) [inline library/std/src/net/ip.rs:1265]: Now there's the opportunity for further simplification, analog to `is_link_local` above:
```suggestion
       matches!(self.segments(), [0xfe80, 0, 0, 0, ..])
```

File: library/std/src/net/ip.rs [hunk 2]
```diff
@@ -1262,10 +1259,7 @@ impl Ipv6Addr {
     /// [RFC 4291 errata 4406]: https://www.rfc-editor.org/errata/eid4406
     #[rustc_const_unstable(feature = "const_ipv6", issue = "76205")]
     pub const fn is_unicast_link_local_strict(&self) -> bool {
-        (self.segments()[0] & 0xffff) == 0xfe80
-            && (self.segments()[1] & 0xffff) == 0
-            && (self.segments()[2] & 0xffff) == 0
-            && (self.segments()[3] & 0xffff) == 0
+        matches!(self.segments(), [0xfe80, 0, 0, 0, ..])
```
Comment (shepmaster) [inline library/std/src/net/ip.rs:1262]: I'd think `starts_with` is clearer:

```rust
fn x(a: std::net::Ipv6Addr) {
    a.segments().starts_with(&[0xfe80, 0, 0, 0]);
}
```

File: library/std/src/net/addr.rs [hunk 3]
```diff
@@ -364,7 +364,6 @@ impl SocketAddrV6 {
                 sin6_addr: *ip.as_inner(),
                 sin6_flowinfo: flowinfo,
                 sin6_scope_id: scope_id,
-                ..unsafe { mem::zeroed() }
```
Comment (shepmaster) [inline library/std/src/net/addr.rs:367]: This is surprising. Which Clippy lint was this? Could there be a platform where there are other fields?

Comment (shepmaster) [reply library/std/src/net/addr.rs:367]: Ah, `needless_update`. The other question remains.

File: library/std/src/net/ip.rs (hunk 2 above)
Comment (wcampbell0x2a) [reply library/std/src/net/ip.rs:1262]: applying this patch gives an error:
```
diff --git a/library/std/src/net/ip.rs b/library/std/src/net/ip.rs
index 8089d7a8ba6..37ccd3618b3 100644
--- a/library/std/src/net/ip.rs
+++ b/library/std/src/net/ip.rs
@@ -1259,7 +1259,7 @@ impl Ipv6Addr {
     /// [RFC 4291 errata 4406]: https://www.rfc-editor.org/errata/eid4406
     #[rustc_const_unstable(feature = "const_ipv6", issue = "76205")]
     pub const fn is_unicast_link_local_strict(&self) -> bool {
-        matches!(self.segments(), [0xfe80, 0, 0, 0, ..])
+        self.segments().starts_with(&[0xfe80, 0, 0, 0])
     }

     /// Returns [`true`] if the address is a unicast link-local address (`fe80::/10`).
```

```
error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
    --> library/std/src/net/ip.rs:1262:9
     |
1262 |         self.segments().starts_with(&[0xfe80, 0, 0, 0])
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

Comment (shepmaster) [reply library/std/src/net/ip.rs:1262]: Ah, I missed this was a `const fn`

File: library/std/src/net/addr.rs (hunk 3 above)
Comment (LingMan) [reply library/std/src/net/addr.rs:367]: Took the liberty to do some digging. `sockaddr_in6` is defined in-tree for some platforms and in the libc crate for others. The in-tree ones all have the fields listed here but libc has more variety. BSD, Haiku, and a few others have an additional `sin6_len: u8` field. Solaris and Illumos have `__sin6_src_id: u32` instead.

Comment (wcampbell0x2a) [reply library/std/src/net/addr.rs:367]: I'm assuming feature gating this would be messy, and I should just revert that change?

Comment (scottmcm) [reply library/std/src/net/addr.rs:367]: Reverting it seems reasonable.  Then the other changes can go in easily.

Comment (wcampbell0x2a) [reply library/std/src/net/addr.rs]: done

**Comment by scottmcm** [comment]
Now that this is just changing some `true`/`false` `match`es to `matches!`, LGTM.

r? @scottmcm 
@bors r+ rollup

**Pull request merged**